### PR TITLE
[WIP] Added Stacks Using Linked List

### PR DESCRIPTION
**Pull request by sarthakforwet**
#### References to other Issues or PRs or Relevant literature
Issue #113 

#### Brief description of what is fixed or changed
I have added linked list implementation of Stacks and working on Linked list implementation of queues as well .

### CODE REVIEW

**Comment by czgdp1807** [comment]
Please allow edits from maintainers. See the check box on the right side of the page. Just tick it.

**Comment by sarthakforwet** [comment]
@czgdp1807 How to know on what basis and what is being checked ?
Thank you

**Comment by czgdp1807** [comment]
@sarthakforwet 
I have removed `build` and `dist` files from your branch, `developer`. Please do the following before pushing any commit,

1. `git remote add origin_user https://github.com/sarthakforwet/pydatastructs`
2. `git pull origin_user developer`
3. `sudo python3 setup.py develop`(if you are on linux or something equivalent for windows).
4. Add your commits
5. Push to your branch.
Let me know if you face any problems.

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -12,7 +12,7 @@ appearance, race, religion, or sexual identity and orientation.
 ## Our Standards
 
 Examples of behavior that contributes to creating a positive environment
-include:
```
Comment (czgdp1807) [inline CODE_OF_CONDUCT.md]: Don't make changes to this file.

File: pydatastructs/miscellaneous_data_structures/linked_stacks.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+        return self
+
+    def pop(self):
+        if self.top is not None:
+            self.data = self.top.data
+            self.top = self.top.next
+        return self.data
+    
+    def peek(self):
+        return self.top.data
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/linked_stacks.py]: Add a new line.

File: pydatastructs/miscellaneous_data_structures/linked_stacks.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from pydatastructs import SinglyLinkedList
+
+class Linked_Stacks:
+    
+    # A class to implement Stacks Built over Singly Linked Lists
+    
+    def __init__(self):
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/linked_stacks.py]: Use a `__new__` method.

File: pydatastructs/miscellaneous_data_structures/linked_stacks.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from pydatastructs import SinglyLinkedList
+
+class Linked_Stacks:
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/linked_stacks.py]: A better name is `LinkedListStack`.

**Comment by czgdp1807** [comment]
Please shift the code to `stacks.py` in `miscellaneous_data_structures`.

**Comment by sarthakforwet** [comment]
Okk I understood the points and also shifted my code to `stacks.py` file .

**Comment by sarthakforwet** [comment]
@czgdp1807 Changes have been made and sucessful so should I close this issue now?

**Comment by czgdp1807** [comment]
Please keep the issue open until the PR is merged.

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (czgdp1807) [reply CODE_OF_CONDUCT.md]: Please address this.

File: build/lib/pydatastructs/miscellaneous_data_structures/stack.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    def peek(self):
+        return self.items[self.items._last_pos_filled]
+
+    def __str__(self):
+        """
+        Used for printing.
+        """
+        return str(self.items._data)
+
+class Linked_Stacks(Stack):
```
Comment (czgdp1807) [inline build/lib/pydatastructs/miscellaneous_data_structures/stack.py]:
```suggestion
class LinkedListStack(Stack):
```

File: build/lib/pydatastructs/miscellaneous_data_structures/stack.py [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+    
+    """Representation of Stack Data Structure using Doubly Linked List
```
Comment (czgdp1807) [inline build/lib/pydatastructs/miscellaneous_data_structures/stack.py]: IMO, we can use `SinglyLinkedList` for stacks.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 7]
```diff
@@ -109,3 +115,68 @@ def __str__(self):
...
+    peek :
+    returns the value of top element
+
+    is_empty :
+    Checks for whether a Stack is been empty or not
+    Return True if empty else False
+
+    __str__ :
+    Used for Printing the Stack
+
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]:
```suggestion
```

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 8]
```diff
@@ -109,3 +115,68 @@ def __str__(self):
...
+            self.top = self.top.prev
+            return self.data
+        else:
+            return "Stack is empty"
+
+    @property
+    def is_empty(self):
+        if self.top is None:
+            return 1
+        return 0
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]:
```suggestion
        return self.top is None
```

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 9]
```diff
@@ -109,3 +115,68 @@ def __str__(self):
...
+
+    @property
+    def peek(self):
+        if self.top is not None:
+            return self.top.data
+        return "Stack is empty"
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: Instead raise a `ValueError`.

File: pydatastructs/miscellaneous_data_structures/stack.py [hunk 10]
```diff
@@ -109,3 +115,68 @@ def __str__(self):
...
+
+    def __str__(self):
+        "Used for Printing the Stack"
+        iterator = self.top
+        while iterator is not None:
+            print(iterator.data)
+            iterator = iterator.next
```
Comment (czgdp1807) [inline pydatastructs/miscellaneous_data_structures/stack.py]: Avoid print statements. Instead return a string.

File: build/lib/pydatastructs/miscellaneous_data_structures/stack.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+            self.top = self.dll.head
+        self.top = self.top.next
+
+    def pop(self):
+        if not self.is_empty:
+            self.data = self.top.data
+            self.top = self.top.prev
+            return self.data
+        else:
+            raise ValueError
```
Comment (manaswinidas) [inline build/lib/pydatastructs/miscellaneous_data_structures/stack.py]: ValueError takes a string argument: https://kite.com/python/answers/how-to-raise-a-valueerror-in-python

File: build/lib/pydatastructs/miscellaneous_data_structures/stack.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+            return 1
+        return 0
+        return self.top is None
+
+
+    @property
+    def peek(self):
+        if self.top is not None:
+            return self.top.data
+        raise ValueError
```
Comment (manaswinidas) [inline build/lib/pydatastructs/miscellaneous_data_structures/stack.py]: ValueError takes a string argument. For your reference: https://kite.com/python/answers/how-to-raise-a-valueerror-in-python

**Comment by czgdp1807** [comment]
Closing in favour of #152

File: build/lib/pydatastructs/miscellaneous_data_structures/stack.py (hunk 11 above)
Comment (sarthakforwet) [reply build/lib/pydatastructs/miscellaneous_data_structures/stack.py]: Not necessary though.

**Pull request closed without merging**